external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriter.scala

```diff
@@ -0,0 +1,264 @@
+  val DEFAULT_WAIT_FOR_CONFIRMED_WRITE_MS = 1000
+
+  private case class TaskCommitMessage(
+    sparkStageId: Int,
+    sparkPartitionId: Int,
+    writeCommitted: Boolean) extends Serializable
+
+  def write(sparkSession: SparkSession,
+    queryExecution: QueryExecution,
+    kafkaParameters: ju.Map[String, Object],
+    defaultTopic: Option[String] = None): Unit = {
+
+    val schema = queryExecution.logical.output
+    schema.find(p => p.name == TOPIC_ATTRIBUTE_NAME).getOrElse(
+      if (defaultTopic == None) {
+        throw new IllegalArgumentException(s"Default topic required when no " +
+          s"'$TOPIC_ATTRIBUTE_NAME' attribute is present")
+      } else {
+        Literal(defaultTopic.get, StringType)
+      }
+    ).dataType match {
+      case StringType => // good
+      case _ =>
+        throw new IllegalArgumentException(s"Topic type must be a String")
+    }
+    schema.find(p => p.name == KEY_ATTRIBUTE_NAME).getOrElse(
+      Literal(null, StringType)
+    ).dataType match {
+      case StringType | BinaryType => // good
+      case _ =>
+        throw new IllegalArgumentException(s"$KEY_ATTRIBUTE_NAME attribute type " +
+          s"must be a String or BinaryType")
+    }
+    val valueField = schema.find(p => p.name == VALUE_ATTRIBUTE_NAME).getOrElse(
+      throw new IllegalArgumentException(s"Required attribute '$VALUE_ATTRIBUTE_NAME' not found")
+    ).dataType match {
+      case StringType | BinaryType => // good
+      case _ =>
+        throw new IllegalArgumentException(s"$VALUE_ATTRIBUTE_NAME attribute type " +
+          s"must be a String or BinaryType")
+    }
+    SQLExecution.withNewExecutionId(sparkSession, queryExecution) {
+      try {
+        val ret = sparkSession.sparkContext.runJob(queryExecution.toRdd,
+          (taskContext: TaskContext, iter: Iterator[InternalRow]) => {
+            executeTask(
+              iterator = iter,
+              producerConfiguration = kafkaParameters,
+              sparkStageId = taskContext.stageId(),
+              sparkPartitionId = taskContext.partitionId(),
+              sparkAttemptNumber = taskContext.attemptNumber(),
+              inputSchema = schema,
+              defaultTopic = defaultTopic)
+          })
+
+        // logInfo(s"Job ${job.getJobID} committed.")
+      } catch {
+        case cause: Throwable =>
+          // logError(s"Aborting job ${job.getJobID}.", cause)
+          throw new SparkException("Job aborted.", cause)
+      }
+    }
+  }
+
+  /** Writes data out in a single Spark task. */
+  private def executeTask(
+      iterator: Iterator[InternalRow],
+      producerConfiguration: ju.Map[String, Object],
+      sparkStageId: Int,
+      sparkPartitionId: Int,
+      sparkAttemptNumber: Int,
+      inputSchema: Seq[Attribute],
+      defaultTopic: Option[String]): TaskCommitMessage = {
+    val writeTask = new KafkaWriteTask(
+      producerConfiguration, inputSchema, defaultTopic)
+    try {
```

Comment:
dont need this additional try when you are using `Utils.tryWithSafeFinallyAndFailureCallbacks`